File: src/java/org/orbeon/oxf/util/PooledXPathExpression.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.util;

import org.apache.commons.pool.ObjectPool;
import org.orbeon.oxf.common.OXFException;
import org.orbeon.oxf.xforms.XFormsUtils;
import org.orbeon.saxon.expr.Expression;
import org.orbeon.saxon.expr.XPathContext;
import org.orbeon.saxon.expr.XPathContextMajor;
import org.orbeon.saxon.om.Item;
import org.orbeon.saxon.om.SequenceIterator;
import org.orbeon.saxon.om.ValueRepresentation;
import org.orbeon.saxon.om.VirtualNode;
import org.orbeon.saxon.sxpath.XPathDynamicContext;
import org.orbeon.saxon.sxpath.XPathExpression;
import org.orbeon.saxon.sxpath.XPathVariable;
import org.orbeon.saxon.trans.XPathException;
import org.orbeon.saxon.value.AtomicValue;
import org.orbeon.saxon.value.SequenceExtent;
import org.orbeon.saxon.value.Value;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class PooledXPathExpression {

    private XPathExpression  expression;
    private ObjectPool pool;
    private Map<String, XPathVariable> variables;

    // Dynamic context
    private Map<String, ValueRepresentation> variableToValueMap;
    private Item contextItem;
    private int contextPosition;

    public PooledXPathExpression(XPathExpression expression, ObjectPool pool, Map<String, XPathVariable> variables) {
        this.expression = expression;
        this.pool = pool;
        this.variables = variables;
    }

    /**
     * This *must* be called in a finally block to return the expression to the pool.
     */
    public void returnToPool() {
        try {
            // Free up dynamic context references
            variableToValueMap = null;
            contextItem = null;

            // Return object to pool
            if (pool != null) // may be null for testing
                pool.returnObject(this);
        } catch (Exception e) {
            throw new OXFException(e);
        }
    }

    public Expression getExpression() {
        return expression.getInternalExpression();
    }

    private Item getContextItem() {
        return contextItem;
    }

    /**
     * Evaluate and return an iterator over native Java objects, including underlying wrapped nodes.
     */
    public Iterator iterate() throws XPathException {

        final Item contextItem = getContextItem();
        final SequenceIterator iter = expression.iterate(expression.createDynamicContext(contextItem));

        return new Iterator() {

            private Item currentItem = iter.next();

            public boolean hasNext() {
                return currentItem != null;
            }

            public Object next() {
                final Item itemToReturn = currentItem;
                // Advance
                try {
                    currentItem = iter.next();
                } catch (XPathException e) {
                    throw new OXFException(e);
                }
                // Convert
                if (itemToReturn instanceof AtomicValue) {
                    try {
                        return Value.convertToJava(itemToReturn);
                    } catch (XPathException e) {
                        throw new OXFException(e);
                    }
                } else if (itemToReturn instanceof VirtualNode) {
                    return ((VirtualNode)itemToReturn).getUnderlyingNode();
                } else {
                    return itemToReturn;
                }
            }

            public void remove() {
                throw new UnsupportedOperationException();
            }
        };
    }

    /**
     * Evaluate and return a List of native Java objects, including underlying wrapped nodes.
     */
    public List evaluate() throws XPathException {
        return convertToJava(evaluate(null));
    }

    /**
     * Evaluate and return a List of native Java objects, but keep NodeInfo objects.
     */
    public List<Object> evaluateKeepNodeInfo(Object functionContext) throws XPathException {
        final SequenceIterator iter = evaluate(functionContext);

        final SequenceExtent extent = new SequenceExtent(iter);
        return convertToJavaKeepNodeInfo(extent);
    }

    /**
     * Evaluate and return a List of Item objects.
     */
    public List<Item> evaluateKeepItems(Object functionContext) throws XPathException {
        final SequenceIterator iter = evaluate(functionContext);

        final List<Item> result = new ArrayList<Item>();
        Item next;
        while ((next = iter.next()) != null) {
            result.add(next);
        }
        return result;
    }

    /**
     * Evaluate and return a List of Item objects.
     */
    public Item evaluateSingleKeepItems(Object functionContext) throws XPathException {
        final SequenceIterator iter = evaluate(functionContext);
        return iter.next();
    }

    /**
     * Evaluate the expression as a variable value usable by Saxon in further XPath expressions.
     */
    public SequenceExtent evaluateAsExtent(Object functionContext) throws XPathException {
        final SequenceIterator iter = evaluate(functionContext);

        return new SequenceExtent(iter);
    }

    private List<Object> convertToJavaKeepNodeInfo(SequenceExtent extent) throws XPathException {
        final List<Object> result = new ArrayList<Object>(extent.getLength());
        final SequenceIterator iter = extent.iterate();
        while (true) {
            final Item currentItem = iter.next();
            if (currentItem == null) {
                return result;
            }
            if (currentItem instanceof AtomicValue) {
                result.add(Value.convertToJava(currentItem));
            } else {
                result.add(currentItem);
            }
        }
    }

//    private List<Object> convertToJava(SequenceExtent extent) throws XPathException {
//        final List<Object> result = new ArrayList<Object>(extent.getLength());
//        final SequenceIterator iter = extent.iterate(null);
//        while (true) {
//            final Item currentItem = iter.next();
//            if (currentItem == null) {
//                return result;
//            }
//            result.add(Value.convertToJava(currentItem));
//        }
//    }

    private List<Object> convertToJava(SequenceIterator iterator) throws XPathException {
        final List<Object> result = new ArrayList<Object>();
        while (true) {
            final Item currentItem = iterator.next();
            if (currentItem == null) {
                return result;
            }
            result.add(Value.convertToJava(currentItem));
        }
    }

    /**
     * Evaluate and return a single native Java object, including underlying wrapped nodes. Return null if the
     * evaluation doesn't return any item.
     */
    public Object evaluateSingle() throws XPathException {
        final SequenceIterator iter = evaluate(null);

        final Item firstItem = iter.next();
        if (firstItem == null) {
            return null;
        } else {
            return Value.convertToJava(firstItem);
        }
    }

    /**
     * Evaluate and return a single native Java object, but keep NodeInfo objects. Return null if the evaluation
     * doesn't return any item.
     */
    public Object evaluateSingleKeepNodeInfo(Object functionContext) throws XPathException {
        final SequenceIterator iter = evaluate(functionContext);

        final Item firstItem = iter.next();
        if (firstItem == null) {
            return null;
        } else {
            if (firstItem instanceof AtomicValue) {
                return Value.convertToJava(firstItem);
            } else {
                return firstItem;
            }
        }
    }

    public Expression prepareExpression(XPathContextMajor xpathContext, Object functionContext) throws XPathException {
        prepareDynamicContext(xpathContext, functionContext);
        return expression.getInternalExpression();
    }

    /**
     * Return the function context passed to the expression if any.
     */
    public static Object getFunctionContext(XPathContext xpathContext) {
        return xpathContext.getController().getUserData("", PooledXPathExpression.class.getName());
    }

    private SequenceIterator evaluate(Object functionContext) throws XPathException {
        final XPathDynamicContext dynamicContext = prepareDynamicContext(null, functionContext);
        return expression.iterate(dynamicContext);
    }

    public XPathDynamicContext prepareDynamicContext(XPathContextMajor xpathContext, Object functionContext) throws XPathException {

        final XPathDynamicContext dynamicContext;
        if (xpathContext == null) {
            dynamicContext = expression.createDynamicContext(contextItem, contextPosition);
            xpathContext = (XPathContextMajor) dynamicContext.getXPathContextObject();
        } else {
            dynamicContext = expression.createDynamicContext(xpathContext, contextItem, contextPosition);
        }

        // Pass function context to controller
        xpathContext.getController().setUserData("", this.getClass().getName(), functionContext);

        // Set variable values if any
        if (variableToValueMap != null) {
            for (Map.Entry<String, XPathVariable> entry: variables.entrySet()) {
                final String name = entry.getKey();
                final XPathVariable variable = entry.getValue();

                final ValueRepresentation object = variableToValueMap.get(name);
                if (object != null) {
                    // Convert Java object to Saxon object
                    final ValueRepresentation valueRepresentation = XFormsUtils.convertJavaObjectToSaxonObject(object);
                    xpathContext.setLocalVariable(variable.getLocalSlotNumber(), valueRepresentation);
                }
            }
        }

        return dynamicContext;
    }

    /**
     * Set context node-set and initial position.
     *
     * @param contextItems          List of Item
     * @param contextPosition       1-based current position
     */
    public void setContextItems(List<Item> contextItems, int contextPosition) {

        if (contextPosition > 0 && contextPosition <= contextItems.size())
            setContextItem(contextItems.get(contextPosition - 1), contextPosition);
        else
            setContextItem(null, 0);
    }

    public void setContextItem(Item contextItem, int contextPosition) {
        this.contextItem = contextItem;
        this.contextPosition = contextPosition;
    }

    public void setVariables(Map<String, ValueRepresentation> variableToValueMap) {
        // NOTE: We used to attempt to detect whether the expression required variables or not and throw an exception,
        // but we can't really detect this because there may be variables in scope even if the expression does not use
        // them. Conversely, if there are undeclared variables, we let the XPath engine complain about that.
        this.variableToValueMap = variableToValueMap;
    }

    public void destroy() {
        expression = null;
        pool = null;
        variables = null;

        variableToValueMap = null;
        contextItem = null;
    }
}

File: src/main/scala/org/orbeon/oxf/util/PooledXPathExpression.scala
/**
 * Copyright (C) 2013 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.util

import java.{util => ju}

import org.apache.commons.pool.ObjectPool
import org.orbeon.oxf.common.OXFException
import org.orbeon.oxf.util.XPath._
import org.orbeon.saxon.expr.{Expression, XPathContextMajor}
import org.orbeon.saxon.om.{Item, SequenceIterator, ValueRepresentation}
import org.orbeon.saxon.sxpath.{XPathDynamicContext, XPathExpression, XPathVariable}
import org.orbeon.saxon.value.{AtomicValue, ObjectValue, SequenceExtent, Value}
import org.orbeon.scaxon.Implicits

import scala.jdk.CollectionConverters._
import scala.collection.mutable
import scala.util.control.NonFatal
import scala.collection.compat._

class PooledXPathExpression(
  expression : XPathExpression,
  pool       : ObjectPool[PooledXPathExpression],
  variables  : List[(String, XPathVariable)]
) {

  // FIXME: This shouldn't be mutable state
  private var variableToValueMap: ju.Map[String, ValueRepresentation] = null
  private var contextItem: Item = null
  private var contextPosition: Int = 0

  /**
   * Set context node-set and initial position.
   *
   * @param contextItems          List of Item
   * @param contextPosition       1-based current position
   */
  def setContextItems(contextItems: ju.List[Item], contextPosition: Int): Unit =
    if (contextPosition > 0 && contextPosition <= contextItems.size)
      setContextItem(contextItems.get(contextPosition - 1), contextPosition)
    else
      setContextItem(null, 0)

  def setContextItem(contextItem: Item, contextPosition: Int): Unit = {
    this.contextItem = contextItem
    this.contextPosition = contextPosition
  }

  def setVariables(variableToValueMap: ju.Map[String, ValueRepresentation]): Unit =
    this.variableToValueMap = variableToValueMap

  /**
   * This *must* be called in a finally block to return the expression to the pool.
   */
  def returnToPool(): Unit = {
    variableToValueMap = null
    contextItem = null
    if (pool ne null)
      pool.returnObject(this)
  }

  def internalExpression: Expression = expression.getInternalExpression

  /**
   * Evaluate and return a List of native Java objects, but keep NodeInfo objects.
   *
   * NOTE: Used by XPathCache.
   */
  def evaluateKeepNodeInfo(functionContext: FunctionContext): ju.List[Any] =
    withFunctionContext(functionContext) {
      scalaIteratorToJavaList(Implicits.asScalaIterator(evaluate()) map itemToJavaKeepNodeInfoOrNull)
    }

  /**
   * Evaluate and return a List of Item objects.
   *
   * NOTE: Used by XPathCache.
   */
  def evaluateKeepItemsJava(functionContext: FunctionContext): ju.List[Item] =
    withFunctionContext(functionContext) {
      scalaIteratorToJavaList(Implicits.asScalaIterator(evaluate()))
    }

  // NOTE: Used by XPathCache.
  def evaluateKeepItems(functionContext: FunctionContext): List[Item] =
    withFunctionContext(functionContext) {
      Implicits.asScalaIterator(evaluate()).toList
    }

  /**
   * Evaluate and return a List of Item objects.
   *
   * NOTE: Used by XPathCache.
   */
  def evaluateSingleKeepItemOrNull(functionContext: FunctionContext): Item =
    withFunctionContext(functionContext) {
      evaluate().next()
    }

  /**
   * Evaluate the expression as a variable value usable by Saxon in further XPath expressions.
   *
   * NOTE: Used by XPathCache.
   */
  def evaluateAsExtent(functionContext: FunctionContext): SequenceExtent =
    withFunctionContext(functionContext) {
      new SequenceExtent(evaluate())
    }

  /**
   * Evaluate and return a single native Java object, but keep NodeInfo objects. Return null if the evaluation
   * doesn't return any item.
   *
   * NOTE: Used by XPathCache.
   */
  def evaluateSingleKeepNodeInfoOrNull(functionContext: FunctionContext): AnyRef =
    withFunctionContext(functionContext) {
      singleItemToJavaKeepNodeInfoOrNull(evaluate().next())
    }

  private def scalaIteratorToJavaList[T](i: Iterator[T]): ju.List[T] =
    new ju.ArrayList(i.to(mutable.ArrayBuffer).asJava)

  private def itemToJavaKeepNodeInfoOrNull(item: Item) = item match {
    case v: ObjectValue => v // don't convert for `Array` and `Map` types
    case v: AtomicValue => Value.convertToJava(v)
    case v              => v
  }

  private def singleItemToJavaOrNull(item: Item) = item match {
    case null => null
    case item => Value.convertToJava(item)
  }

  private def singleItemToJavaKeepNodeInfoOrNull(item: Item) = item match {
    case null => null
    case item => itemToJavaKeepNodeInfoOrNull(item)
  }

  /**
   * Evaluate and return an iterator over native Java objects, including underlying wrapped nodes.
   *
   * NOTE: Used by legacy Java code only.
   */
  def iterateReturnJavaObjects: ju.Iterator[AnyRef] =
    Implicits.asScalaIterator(expression.iterate(expression.createDynamicContext(contextItem))) map Value.convertToJava asJava

  /**
   * Evaluate and return a List of native Java objects, including underlying wrapped nodes.
   *
   * NOTE: Used by legacy Java code only.
   */
  def evaluateToJavaReturnToPool: ju.List[AnyRef] =
    try scalaIteratorToJavaList(Implicits.asScalaIterator(evaluate()) map Value.convertToJava)
    finally returnToPool()

  /**
   * Evaluate and return a single native Java object, including underlying wrapped nodes. Return null if the
   * evaluation doesn't return any item.
   *
   * NOTE: Used by legacy Java code only.
   */
  def evaluateSingleToJavaReturnToPoolOrNull: AnyRef =
    try singleItemToJavaOrNull(evaluate().next())
    catch { case NonFatal(e) => throw new OXFException(e) } // so Java callers can catch a RuntimeException
    finally returnToPool()

  private def evaluate(): SequenceIterator = {
    val (dynamicContext, xpathContext) = newDynamicAndMajorContexts
    prepareDynamicContext(xpathContext)
    expression.iterate(dynamicContext)
  }

  // Called from exf:sort(), indirectly from xxf:evaluate-avt(), and evaluate()
  def prepareDynamicContext(xpathContext: XPathContextMajor): Unit =
    if (variableToValueMap ne null) {
      for ((name, variable) <- variables) {
        val value = variableToValueMap.get(name)
        if (value ne null) // FIXME: this should never happen, right?
          xpathContext.setLocalVariable(variable.getLocalSlotNumber, value)
      }
    }

  // Called from `exf:sort()` and `evaluate()`
  def newDynamicAndMajorContexts: (XPathDynamicContext, XPathContextMajor) = {
    val dynamicContext = expression.createDynamicContext(contextItem, contextPosition)
    (dynamicContext, dynamicContext.getXPathContextObject.asInstanceOf[XPathContextMajor])
  }
}
